src/clustering/image_clustering.h

```diff
@@ -38,36 +38,85 @@
 #include <queue>
 #include <unordered_map>
 #include <vector>
+#include <mutex>
 
 #include "clustering/cluster.h"
 #include "graph/graph.h"
 #include "util/hash.h"
 #include "util/timer.h"
 #include "util/types.h"
+#include "util/threading.h"
 
 using namespace colmap;
 
 namespace DAGSfM {
 
 using Edges = std::unordered_map<ImagePair, int>;
-struct ImageCluster {
+class ImageCluster {
+ private:
+  ImageCluster(const ImageCluster& rhs, const std::lock_guard<std::mutex> &) 
+  : cluster_id(rhs.cluster_id), image_ids_(rhs.image_ids_), edges_(rhs.edges_),
+    is_condition_satisfy_(rhs.is_condition_satisfy_), completed_(rhs.completed_) {}
+
+ public:
+  ImageCluster() = default;
+  // Thread safe copy constructor, and copy assignment operator.
+  ImageCluster(const ImageCluster& rhs) : ImageCluster(rhs, std::lock_guard<std::mutex>(rhs.mutex_)) {}
+  ImageCluster& operator=(const ImageCluster& rhs) {
+    if (this != &rhs) {
+      std::unique_lock<std::mutex> my_lock(mutex_, std::defer_lock);
+      std::unique_lock<std::mutex> rhs_lock(rhs.mutex_, std::defer_lock);
+      std::lock(my_lock, rhs_lock);
+      cluster_id = rhs.cluster_id;
+      image_ids_ = rhs.image_ids_;
+      edges_ = rhs.edges_;
+      is_condition_satisfy_ = rhs.is_condition_satisfy_;
+      completed_ = rhs.completed_;
+    }
+    return *this;
+  }
+ public:
   int cluster_id = 0;
```

Comment:
To keep the code's consistency and obey the data capsulation rule, `cluster_id` should also be a private member variable, and a setter and a getter should also be provided as public functions.